File: backend/websocket/manager.py
import json
from fastapi import WebSocket


class ConnectionManager:
    def __init__(self):
        # user_id -> WebSocket
        self.active_connections: dict[str, WebSocket] = {}
        # user_id -> set of conversation_ids
        self.user_rooms: dict[str, set[str]] = {}

    async def connect(self, websocket: WebSocket, user_id: str):
        self.active_connections[user_id] = websocket
        self.user_rooms[user_id] = set()

        # Send currently online users to the new connection
        for online_id in self.active_connections:
            if online_id != user_id:
                await self.send_to_user(user_id, {
                    "type": "user:online",
                    "userId": online_id,
                })

        # Notify others
        await self.broadcast({
            "type": "user:online",
            "userId": user_id,
        }, exclude=user_id)

    async def disconnect(self, user_id: str):
        self.active_connections.pop(user_id, None)
        self.user_rooms.pop(user_id, None)

        await self.broadcast({
            "type": "user:offline",
            "userId": user_id,
        })

    def join_room(self, user_id: str, conversation_id: str):
        if user_id in self.user_rooms:
            self.user_rooms[user_id].add(conversation_id)

    async def send_to_user(self, user_id: str, message: dict):
        ws = self.active_connections.get(user_id)
        if ws:
            try:
                await ws.send_json(message)
            except Exception:
                pass

    async def broadcast(self, message: dict, exclude: str | None = None):
        for uid, ws in list(self.active_connections.items()):
            if uid != exclude:
                try:
                    await ws.send_json(message)
                except Exception:
                    pass

    async def broadcast_to_room(self, conversation_id: str, message: dict, exclude: str | None = None):
        for uid, rooms in self.user_rooms.items():
            if conversation_id in rooms and uid != exclude:
                await self.send_to_user(uid, message)

    def get_online_users(self) -> dict[str, WebSocket]:
        return self.active_connections

    def is_online(self, user_id: str) -> bool:
        return user_id in self.active_connections


manager = ConnectionManager()

File: src/lib/ws/manager.ts
import type { WebSocket } from "ws";

class ConnectionManager {
  // Multiple sockets per userId to support multi-session (e.g. localhost + network IP)
  private connections = new Map<string, Set<WebSocket>>();
  private rooms = new Map<string, Set<string>>();

  async connect(ws: WebSocket, userId: string): Promise<void> {
    const isFirstSession = !this.connections.has(userId);

    if (isFirstSession) {
      this.connections.set(userId, new Set());
      this.rooms.set(userId, new Set());
    }
    this.connections.get(userId)!.add(ws);

    // Tell the new socket about everyone already online
    for (const onlineId of this.connections.keys()) {
      if (onlineId !== userId) {
        this.sendToSocket(ws, { type: "user:online", userId: onlineId });
      }
    }

    // Notify others only when the first session for this user connects
    if (isFirstSession) {
      this.broadcast({ type: "user:online", userId }, userId);
    }
  }

  disconnect(userId: string, ws: WebSocket): void {
    const sockets = this.connections.get(userId);
    if (!sockets) return;

    sockets.delete(ws);

    if (sockets.size === 0) {
      // Last session gone — user is now offline
      this.connections.delete(userId);
      this.rooms.delete(userId);
      this.broadcast({ type: "user:offline", userId });
    }
  }

  joinRoom(userId: string, conversationId: string): void {
    this.rooms.get(userId)?.add(conversationId);
  }

  sendToUser(userId: string, message: Record<string, unknown>): void {
    const sockets = this.connections.get(userId);
    if (!sockets) return;
    for (const ws of sockets) {
      this.sendToSocket(ws, message);
    }
  }

  private sendToSocket(ws: WebSocket, message: Record<string, unknown>): void {
    try {
      ws.send(JSON.stringify(message));
    } catch {
      // Drop on error; the close handler will clean up
    }
  }

  broadcast(message: Record<string, unknown>, excludeUserId?: string): void {
    for (const [uid] of this.connections) {
      if (uid !== excludeUserId) this.sendToUser(uid, message);
    }
  }

  broadcastToRoom(
    conversationId: string,
    message: Record<string, unknown>,
    excludeUserId?: string
  ): void {
    for (const [uid, rooms] of this.rooms) {
      if (rooms.has(conversationId) && uid !== excludeUserId) {
        this.sendToUser(uid, message);
      }
    }
  }

  getOnlineUserIds(): string[] {
    return [...this.connections.keys()];
  }

  isOnline(userId: string): boolean {
    return (this.connections.get(userId)?.size ?? 0) > 0;
  }
}

declare global {
  // eslint-disable-next-line no-var
  var _wsManager: ConnectionManager | undefined;
}

export const manager: ConnectionManager =
  global._wsManager ?? (global._wsManager = new ConnectionManager());
